sycl/include/sycl/multi_ptr.hpp

```diff
@@ -27,7 +71,578 @@ template <typename dataT, int dimensions> class local_accessor;
 /// qualified pointers to allow interoperability between plain C++ and OpenCL C.
 ///
 /// \ingroup sycl_api
-template <typename ElementType, access::address_space Space> class multi_ptr {
+// TODO: Default value for DecorateAddress is for backwards compatiblity. It
+//       should be removed.
+template <typename ElementType, access::address_space Space,
+          access::decorated DecorateAddress = access::decorated::legacy>
+class multi_ptr {
+private:
+  using decorated_type =
+      typename detail::DecoratedType<ElementType, Space>::type;
+
+public:
+  static constexpr bool is_decorated =
+      DecorateAddress == access::decorated::yes;
+  static constexpr access::address_space address_space = Space;
+
+  using value_type = ElementType;
+  using pointer = std::conditional_t<is_decorated, decorated_type *,
+                                     std::add_pointer_t<value_type>>;
+  using reference = std::conditional_t<is_decorated, decorated_type &,
+                                       std::add_lvalue_reference_t<value_type>>;
+  using iterator_category = std::random_access_iterator_tag;
+  using difference_type = std::ptrdiff_t;
+
+  static_assert(std::is_same<remove_decoration_t<pointer>,
+                             std::add_pointer_t<value_type>>::value);
+  static_assert(std::is_same<remove_decoration_t<reference>,
+                             std::add_lvalue_reference_t<value_type>>::value);
+  // Legacy has a different interface.
+  static_assert(DecorateAddress != access::decorated::legacy);
+  // constant_space is only supported in legacy multi_ptr.
+  static_assert(Space != access::address_space::constant_space,
+                "SYCL 2020 multi_ptr does not support the deprecated "
+                "constant_space address space.");
```

Comment:
Do I understand it correctly that this is a spec violation? I see that there is a constructor, which has no restrictions about creating a `multi_ptr` with constant address space:
```
explicit multi_ptr(typename multi_ptr<ElementType, Space, access::decorated::yes>::pointer);
```

Reply:
That is not completely clear. I have opened an issue about it: https://github.com/KhronosGroup/SYCL-Docs/issues/288

It seems like the intention of the spec is to say that it is deprecated and as such is only supported by the legacy interface, but no conclusion is made yet. Until it is, I would rather we not potentially waste effort getting constant_space to work in the new interface.